src/utils/file.c

```diff
@@ -747,20 +855,124 @@ ssize_t fio_fwrite_compressed(FILE* f, void const* buf, size_t size, int compres
+		char *errormsg = NULL;
+		int32 uncompressed_size = fio_decompress(uncompressed_buf, buf, size, compress_alg, &errormsg);
+
+		if (uncompressed_size < 0)
+			elog(ERROR, "%s", errormsg);
 
-		return (uncompressed_size < 0)
-			? uncompressed_size
-			: fwrite(uncompressed_buf, 1, uncompressed_size, f);
+		return fwrite(uncompressed_buf, 1, uncompressed_size, f);
 	}
 }
 
-static ssize_t
+static void
 fio_write_compressed_impl(int fd, void const* buf, size_t size, int compress_alg)
 {
+	int rc;
+	int32 uncompressed_size;
 	char uncompressed_buf[BLCKSZ];
-	int32 uncompressed_size = fio_decompress(uncompressed_buf, buf, size, compress_alg);
-	return fio_write_all(fd, uncompressed_buf, uncompressed_size);
+
+	/* If the previous command already have failed,
+	 * then there is no point in bashing a head against the wall
+	 */
+	if (async_errormsg)
+		return;
+
+	/* TODO: We cannot allow error out in fio_decompress */
+	uncompressed_size = fio_decompress(uncompressed_buf, buf, size, compress_alg, &async_errormsg);
+
+	if (uncompressed_size < 0)
+		return;
+
+	rc = write(fd, uncompressed_buf, uncompressed_size);
+
+	if (rc <= 0)
+	{
+		async_errormsg = pgut_malloc(ERRMSG_MAX_LEN);
+		snprintf(async_errormsg, ERRMSG_MAX_LEN, "%s", strerror(errno));
+	}
+}
+
+/* check if remote agent encountered any error during execution of async operations */
+int
+fio_check_error_file(FILE* f, char **errmsg)
+{
+	if (fio_is_remote_file(f))
+	{
+		fio_header hdr;
+
+		hdr.cop = FIO_GET_ASYNC_ERROR;
+		hdr.size = 0;
+
+		IO_CHECK(fio_write_all(fio_stdout, &hdr, sizeof(hdr)), sizeof(hdr));
+
+		/* check results */
+		IO_CHECK(fio_read_all(fio_stdin, &hdr, sizeof(hdr)), sizeof(hdr));
+
+		if (hdr.size > 0)
+		{
+			*errmsg = pgut_malloc(ERRMSG_MAX_LEN);
+			IO_CHECK(fio_read_all(fio_stdin, *errmsg, hdr.size), hdr.size);
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/* check if remote agent encountered any error during execution of async operations */
+int
+fio_check_error_fd(int fd, char **errmsg)
+{
+	if (fio_is_remote_fd(fd))
+	{
+		fio_header hdr;
+
+		hdr.cop = FIO_GET_ASYNC_ERROR;
+		hdr.size = 0;
+
+		IO_CHECK(fio_write_all(fio_stdout, &hdr, sizeof(hdr)), sizeof(hdr));
+
+		/* check results */
+		IO_CHECK(fio_read_all(fio_stdin, &hdr, sizeof(hdr)), sizeof(hdr));
+
+		if (hdr.size > 0)
+		{
+			*errmsg = pgut_malloc(ERRMSG_MAX_LEN);
+			IO_CHECK(fio_read_all(fio_stdin, *errmsg, hdr.size), hdr.size);
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static void
+fio_get_async_error_impl(int out)
+{
+	fio_header hdr;
+	hdr.cop = FIO_GET_ASYNC_ERROR;
+
+	/* send error message */
+	if (async_errormsg)
+	{
+		hdr.size = strlen(async_errormsg) + 1;
+
+		/* send header */
+		IO_CHECK(fio_write_all(out, &hdr, sizeof(hdr)), sizeof(hdr));
+
+		/* send message itself */
+		IO_CHECK(fio_write_all(out, async_errormsg, hdr.size), hdr.size);
+
+		//TODO: should we reset the tainted state ?
+//		pg_free(async_errormsg);
+//		async_errormsg = NULL;
```

Comment:
I don't see, where we reset async_errormsg.
How would we handle several consecutive errors?

Reply:
We don`t. Currently it is assumed, that a first error is the last error.